### ISSUE TITLE
Exporting secret keys

### ISSUE THREAD

**Issue by DBLouis** [gpg-rs/gpgme#33]
The following code fails to export secret keys. The result is an empty file. Public keys however seems to be exported correctly.
I ran it using the master branch.

```rust
use gpgme::{
    Context, CreateKeyFlags as Flags, ExportMode, PassphraseRequest, PinentryMode, Protocol,
};
use std::{fs::File, io::Write, iter, path::PathBuf, time::Duration};

fn main() {
    let validity = Duration::from_secs(3600);

    let mut ctx = Context::from_protocol(Protocol::OpenPgp).unwrap();
    ctx.set_offline(true);
    ctx.set_armor(true);
    ctx.set_pinentry_mode(PinentryMode::Loopback).unwrap();
    ctx.set_engine_home_dir("/tmp").unwrap();

    let mut ctx = ctx.set_passphrase_provider(|_: PassphraseRequest, out: &mut dyn Write| {
        out.write_all("password".as_bytes())?;
        Ok(())
    });

    let res = ctx
        .create_key_with_flags("john doe <john@doe.com>", "ed25519", validity, Flags::CERT)
        .unwrap();

    let fpr = res.fingerprint().unwrap();
    let key = ctx.get_key(fpr).unwrap();

    ctx.create_subkey_with_flags(&key, "ed25519", validity, Flags::AUTH)
        .unwrap();
    ctx.create_subkey_with_flags(&key, "cv25519", validity, Flags::ENCR)
        .unwrap();
    ctx.create_subkey_with_flags(&key, "ed25519", validity, Flags::SIGN)
        .unwrap();

    let mut path = PathBuf::from("/tmp");

    path.push(format!("{}.pub.asc", fpr));
    let file = File::create(&path).unwrap();
    ctx.export(iter::once(fpr), ExportMode::empty(), &file)
        .unwrap();

    path.pop();
    path.push(format!("{}.sec.asc", fpr));
    let file = File::create(&path).unwrap();
    ctx.export(iter::once(fpr), ExportMode::SECRET, &file)
        .unwrap();
}
```

**Comment by johnschug** [comment]
I think think this fails because recent(ish) versions of gpg don't seem to allow export of a password protected secret key without user interaction (see [T2324](https://dev.gnupg.org/T2324): gpgme uses `--batch` internally). I'm not sure why it doesn't return an error (might be worth reporting [upstream](https://dev.gnupg.org/)). If the key doesn't need to be password protected you can use CreateKeyFlags::NOPASSWD to create it without one (for example see [tests/export.rs](https://github.com/gpg-rs/gpgme/blob/7fdbe803e52b0a6f4e0da40c88eb0f733001fb15/tests/export.rs#L27-L28)).

**Comment by DBLouis** [comment]
I need the keys to be password protected. I don't get why is it not working because I do have the password callback set and the engine configured to loopback. I know I successfully made this work before in a shell script so it shouldn't be any different.

**Comment by johnschug** [comment]
The command that gpgme actually executes can be seen by setting the `GPGME_DEBUG` environment variable. With `GPGME_DEBUG=9`, the above `Context::export` call produces something like [this](https://gist.github.com/johnschug/b67aee080dab663cdfa7f44ca1efa82a). The command can be seen on [lines 26-49](https://gist.github.com/johnschug/b67aee080dab663cdfa7f44ca1efa82a#file-test_export-log-L26-L49). It appears that a [passphrase is requested](https://gist.github.com/johnschug/b67aee080dab663cdfa7f44ca1efa82a#file-test_export-log-L94-L98) but is [unable to be retrieved](https://gist.github.com/johnschug/b67aee080dab663cdfa7f44ca1efa82a#file-test_export-log-L106-L110). Comparing this to the command used by an earlier [successful call](https://gist.github.com/johnschug/7ebe636af624a954d087d64efbdd249e) to `Context::create_key`, the major difference seems to be in the `--command-fd 7` argument that is passed. After the passphrase is [requested](https://gist.github.com/johnschug/7ebe636af624a954d087d64efbdd249e), the passphrase callback [writes the passphrase](https://gist.github.com/johnschug/7ebe636af624a954d087d64efbdd249e#file-test_create_key-log-L94-L103) to other end (fd 8) of the pipe.

Looking at the source for [`gpgme_op_createkey`](https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gpgme.git;a=blob;f=src/genkey.c;hb=b088d81cefe45c16409085f1560ad1d4c235a696#l388), it appears that `gpgme_op_createkey` (really `createkey_start`) sets a command handler if a passphrase callback is set on the context. I would suggest opening a bug report (or submitting a patch) upstream to request `gpgme_op_export` be modified to do this if secret key export is requested.

**Comment by DBLouis** [comment]
How can I build this crate using a local install of gpgme located in `/usr/local/`?

**Comment by DBLouis** [comment]
It works with the patch I posted there: https://dev.gnupg.org/T5046.

**Comment by johnschug** [comment]
It looks like the patch has been merged upstream and is included in the 1.15 release.
